build-llvm.py

```diff
@@ -0,0 +1,402 @@
+    else:
+        # evaluate its full path with clang to avoid weird issues and check to
+        # see if it will work with '-fuse-ld', which is what cmake will do. Doing
+        # it now prevents a hard error later.
+        ld = os.environ['LD']
+        if "clang" in cc and clang_version(cc) >= 390:
+            ld = shutil.which(ld)
+        if linker_test(cc, ld):
+            print("LD won't work with " + cc +
+                  ", saving you from yourself by ignoring LD value")
+            ld = None
+
+    # Print what binaries we are using to compile/link with so the user can
+    # decide if that is proper or not
+    print("CC: " + cc)
+    print("CXX: " + cxx)
+    if ld is not None:
+        ld = ld.rstrip()
+        ld_to_print = shutil.which("ld." + ld)
+        if ld_to_print is None:
+            ld_to_print = shutil.which(ld)
+        print("LD: " + ld_to_print)
+
+    return cc, cxx, ld
+
+
+# Make sure that the base dependencies of cmake, curl, git, and ninja are installed
+def check_dependencies():
+    utils.header("Checking dependencies")
+    required_commands = ["cmake", "curl", "git", "ninja"]
+    for command in required_commands:
+        output = shutil.which(command)
+        if output is None:
+            raise RuntimeError(command +
+                               " could not be found, please install it!")
+        print(output)
+
+
+# Download llvm and binutils or update them if they exist
+def fetch_llvm_binutils(root, update, branch):
+    p = root.joinpath("llvm-project")
+    if p.is_dir():
+        if update:
+            utils.header("Updating LLVM")
+            subprocess.run(
+                ["git", "-C", p.as_posix(), "checkout", branch], check=True)
+            subprocess.run(
+                ["git", "-C", p.as_posix(), "pull", "--rebase"], check=True)
+    else:
+        utils.header("Downloading LLVM")
+        subprocess.run([
+            "git", "clone", "-b", branch, "git://github.com/llvm/llvm-project",
+            p.as_posix()
+        ],
+                       check=True)
+
+    # One might wonder why we are downloading binutils in an LLVM build script :)
+    # We need it for the LLVMgold plugin, which can be used for LTO with ld.gold,
+    # which at the time of writing this, is how the Google Pixel 3 kernel is built
+    # and linked.
+    utils.download_binutils(root)
+
+
+# Clean up and create the build folder
+def cleanup(build, incremental):
+    if not incremental and build.is_dir():
+        shutil.rmtree(build.as_posix())
+    build.mkdir(parents=True, exist_ok=True)
+
+
+# Invoke cmake to generate the build files
+def invoke_cmake(build, cc, cxx, debug, install_folder, ld, projects, root,
+                 targets):
+    utils.header("Configuring LLVM")
+
+    # Base cmake defintions, which don't depend on any user supplied options
+    defines = {}
+    # Objective-C Automatic Reference Counting (we don't use Objective-C)
+    defines['CLANG_ENABLE_ARCMT'] = 'OFF'
```

Comment:
It would be good to add the link to the page that describes these.

Reply:
0358cf1514858543b2ec5ddee695811150f23036